src/java/org/apache/cassandra/index/sai/disk/format/IndexDescriptor.java

```diff
@@ -0,0 +1,530 @@
+import org.apache.cassandra.index.sai.disk.io.IndexOutputWriter;
+import org.apache.cassandra.index.sai.memory.RowMapping;
+import org.apache.cassandra.index.sai.utils.IndexFileUtils;
+import org.apache.cassandra.index.sai.utils.PrimaryKey;
+import org.apache.cassandra.index.sai.utils.PrimaryKeyFactory;
+import org.apache.cassandra.io.sstable.Component;
+import org.apache.cassandra.io.sstable.Descriptor;
+import org.apache.cassandra.io.sstable.format.SSTableReader;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.io.util.FileHandle;
+import org.apache.cassandra.utils.FBUtilities;
+import org.apache.cassandra.utils.Pair;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.util.IOUtils;
+
+/**
+ * The {@link IndexDescriptor} is an analog of the SSTable {@link Descriptor} and provides version
+ * specific information about the on-disk state of a {@link StorageAttachedIndex}.
+ *
+ * The {@link IndexDescriptor} is primarily responsible for maintaining a view of the on-disk state
+ * of an index for a specific {@link org.apache.cassandra.io.sstable.SSTable}. It maintains mappings
+ * of the current on-disk components and files. It is responsible for opening files for use by
+ * writers and readers.
+ *
+ * It's remaining responsibility is to act as a proxy to the {@link OnDiskFormat} associated with the
+ * index {@link Version}.
+ */
+public class IndexDescriptor
+{
+    private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+
+    public final Version version;
+    public final Descriptor descriptor;
+    public final ClusteringComparator clusteringComparator;
+    public final PrimaryKeyFactory primaryKeyFactory;
+    public final Set<IndexComponent> perSSTableComponents = Sets.newHashSet();
+    public final Map<String, Set<IndexComponent>> perIndexComponents = Maps.newHashMap();
+    public final Map<IndexComponent, File> onDiskPerSSTableFileMap = Maps.newHashMap();
+    public final Map<IndexComponent, File> onDiskPerSSTableTemporaryFileMap = Maps.newHashMap();
+    public final Map<Pair<IndexComponent, String>, File> onDiskPerIndexFileMap = Maps.newHashMap();
+    public final Map<Pair<IndexComponent, String>, File> onDiskPerIndexTemporaryFileMap = Maps.newHashMap();
+
+    private IndexDescriptor(Version version, Descriptor descriptor, ClusteringComparator clusteringComparator)
+    {
+        this.version = version;
+        this.descriptor = descriptor;
+        this.clusteringComparator = clusteringComparator;
+        this.primaryKeyFactory = PrimaryKey.factory(clusteringComparator);
+    }
+
+    public static IndexDescriptor create(Descriptor descriptor, ClusteringComparator clusteringComparator)
+    {
+        return new IndexDescriptor(Version.LATEST, descriptor, clusteringComparator);
+    }
+
+    public static IndexDescriptor create(SSTableReader sstable)
+    {
+        for (Version version : Version.ALL)
```

Comment:
So at the moment, with only one version, I guess it doesn't matter, but what is the intent when there are multiple versions? Wouldn't it be possible to have two versions w/ the same per-SSTable components?

Reply:
The idea is that the `OnDiskFormat` would support an `IndexFeatureSet` that would detail what each version supported. A query would then defer to the minimum viable feature set that was available across all the indexes. 

It would not be possible to have multiple per-SSTable components because the index creation is tied to the SSTable so all the index files associated with the SSTable would be one version or another. 

The `IndexFeatureSet` is meant to apply to index files from different SSTables having different feature sets not the same SSTable.